### DISCUSSION TITLE
How can I use Open ID Connect (a type of authentication) in a react-native-windows app?

### DISCUSSION THREAD

**Discussion by shirakaba** [microsoft/react-native-windows#13538] [Q&A] [answered]
Hi all, I'm trying to use OpenID Connect in my `react-native-windows` app.

For macOS, I'm using `react-native-app-auth`, based on my [PR](<https://github.com/FormidableLabs/react-native-app-auth/pull/1002>) to add macOS support. This was easy enough to implement, as the underlying CocoaPod, [AppAuth-iOS](<https://github.com/openid/AppAuth-iOS>), had a macOS implementation to begin with.

For Windows, however, I'm stumped. I can see that the organisation behind AppAuth also offer a Node.js implementation, [AppAuth-JS](<https://github.com/openid/AppAuth-JS>), which they've demonstrated usage of [with Electron](<https://github.com/googlesamples/appauth-js-electron-sample?tab=readme-ov-file>). However, integrating it with my RNW app is another question.

First, I've tried running their sample Node.js app, only to find that it uses the "loopback" flow, which Google say they've [deprecated](<https://developers.google.com/identity/protocols/oauth2/resources/loopback-migration>) on non-desktop platforms. This is confusing for a start, because I'm experiencing the error despite running on a desktop platform. But either way, they explicitly say that they've deprecated it on UWP, even if they haven't deprecated it on other types of Windows app.

![image](https://github.com/user-attachments/assets/40c08708-3681-42ea-ab6a-c128e2e0a221)

To my understanding, `react-native-windows` produces a UWP app, so I have no auth solution. I can see an alternative library, [IdentityModel.OidcClient](<https://github.com/IdentityModel/IdentityModel.OidcClient>), which has [samples](<https://github.com/IdentityModel/IdentityModel.OidcClient.Samples>) for a bunch of windows platforms I've never heard of, yet also says that they no longer support UWP.

Can anyone advise on how I might implement OpenID in a `react-native-windows` app?

**Comment by marlenecota** [comment, marked as answer]
Here's some documentation on how to do OpenID in UWP:
https://learn.microsoft.com/en-us/windows/uwp/security/web-authentication-broker

**Reply by shirakaba** [reply]
I've begun to have some success building a native module based on the [winrt::Windows::Security::Authentication::Web::WebAuthenticationBroker::AuthenticateAsync](https://learn.microsoft.com/en-us/uwp/api/windows.security.authentication.web.webauthenticationbroker.authenticateasync?view=winrt-26100) API mentioned in those docs you linked, thank you!

I've been referring to:

- the [C++ sample](https://github.com/microsoft/Windows-universal-samples/blob/main/Samples/WebAuthenticationBroker/cpp/Scenario2_oAuthFacebook.xaml.cpp) for Facebook OAuth
- Microsoft's [samples](https://github.com/microsoft/react-native-windows/blob/81a53b820e04c566a56555598fc4a4ecaa223019/packages/sample-apps/windows/SampleLibraryCPP/SampleModuleCPP.h#L79) for writing a C++ native module for React Native Windows

Hoping to be able to opensource my library at some point. Currently it's coupled to my internal app as I've been struggling with setting up standalone RNW libraries, so to anyone else in need of this, all I can do for now is give some guidance.

1. Supposing your app is named `banana`, and your OpenID provider is configured with a callback URL of `banana://callback`, you will need to register for `banana://` deeplinks by making this addition to your `windows/banana/Package.appxmanifest` file:
  ```diff
    <Applications>
      <Application
        Id="App"
        Executable="$targetnametoken$.exe"
        EntryPoint="banana.App">
        <uap:VisualElements
          DisplayName="banana"
          Square150x150Logo="Assets\Square150x150Logo.png"
          Square44x44Logo="Assets\Square44x44Logo.png"
          Description="banana"
          BackgroundColor="transparent">
          <uap:DefaultTile Wide310x150Logo="Assets\Wide310x150Logo.png"/>
          <uap:SplashScreen Image="Assets\SplashScreen.png" />
        </uap:VisualElements>
  +     <Extensions>
  +       <uap:Extension Category="windows.protocol">
  +         <uap:Protocol Name="banana">
  +           <uap:DisplayName>banana</uap:DisplayName>
  +         </uap:Protocol>
  +       </uap:Extension>
  +     </Extensions>
      </Application>
    </Applications>
  ```
2. The `WebAuthenticationBroker::AuthenticateAsync` method must be run on the UI thread. If you attempt to run it on the React Native thread, it won't crash your app, but it will end up always returning `WebAuthenticationStatus::UserCancel` without even attempting to open a browser. To run on the UI thread, I found that you can use [UIDispatcher](https://learn.microsoft.com/en-us/previous-versions/windows/server-essentials/gg514685(v=msdn.10)):
  ```cpp
  REACT_METHOD(authenticate)
  void authenticate(
      std::string authUrl,
      std::string redirectUri,
      std::function<void(std::string)>&& onSuccess,
      std::function<void(std::string)>&& onFailure
  ) noexcept {
      m_reactContext.UIDispatcher().Post([authUrl, redirectUri, onSuccess, onFailure]() {
          // Call WebAuthenticationBroker::AuthenticateAsync() inside here
      });
  }
  ```
